### PR TITLE
Blocks API from POSTing empty msgs to group chat

### PR DESCRIPTION
**Pull request by FriendlyPopo**
Returns 204 on empty or undefined message.

This is step No.1 of issue #4872

### CODE REVIEW

**Comment by Alys** [comment]
I've tested this and it works as described. I've forced a restart of the failed Travis CI build, just to make sure it doesn't pick up anything wrong (the failure was not due to this PR's changes).

**Comment by Alys** [comment]
Grrr. Travis CI build is still failing. I'm going to wait a couple of days because I think related tests are being worked on elsewhere.

File: website/src/controllers/groups.js [hunk 1]
```diff
@@ -236,24 +236,28 @@ api.getChat = function(req, res, next) {
  * TODO make this it's own ngResource so we don't have to send down group data with each chat post
  */
 api.postChat = function(req, res, next) {
-  var user = res.locals.user
-  var group = res.locals.group;
-  if (group.type!='party' && user.flags.chatRevoked) return res.json(401,{err:'Your chat privileges have been revoked.'});
-  var lastClientMsg = req.query.previousMsg;
-  var chatUpdated = (lastClientMsg && group.chat && group.chat[0] && group.chat[0].id !== lastClientMsg) ? true : false;
+  if(typeof req.query.message === 'undefined' || req.query.message == '') {
```
Comment (paglias) [inline website/src/controllers/groups.js:239]: `if(!req.query.message)` cover `undefined`, `''` but also `null` and it's shorter

**Comment by paglias** [comment]
@Horacious thanks for the pr. `204` is not really an error code but a success code when no content is provided in the response. You should use a 4xx error code

**Comment by Alys** [comment]
@Horacious The Travis CI build is failing first because of this test:

 1) API With token and user id Groups Party Chat Posts a message to party chat:
     Uncaught Error: expected 204 to equal 200

I expect the test posts an empty message. The other tests after that fail presumably because the test message doesn't exist.

So, the tests need to be updated before we merge this PR. We'd probably want to have one test that has an empty message and expects 204 (or whatever status code this PR ends up using), and then adjust the existing "Posts a message" test to post text.

Are you interested in doing that? It's fine if you aren't! I suspect that cuddling your new baby is more fun. :)  If you don't want to, I might give it a go.

**Comment by FriendlyPopo** [comment]
@paglias I believe 204 is the appropriate message. The request -is- a success. How the app deals with it is another matter.

@Alys I'm not sure about "more fun", but it certainly is more time-consuming! =)  You can give it a go, I'll try my hand at it when life gets a bit less hectic again.

**Comment by gisikw** [comment]
@Horacious It's not really a success. The system is explicitly rejecting the message (and rightly so) because the client submitted bad information (akin to trying to register an account without a password, for example).

"The 4xx class of status code is intended for cases in which the client seems to have erred."

Contrasted with "204 No Content
The server successfully processed the request, but is not returning any content. Usually used as a response to a successful delete request."

**Comment by FriendlyPopo** [comment]
@gisikw @paglias I stand corrected!

**Comment by crookedneighbor** [comment]
@Horacious @Alys I've looked at the test failures, and I think I figured it out.

It's because I wrote tests that don't actually test the chat :cry: 

We do some funky stuff with the chat, so while it should be using `req.body` to send the message, for some reason we have it sending via `req.query`. You can see comments about it here: https://github.com/HabitRPG/habitrpg/blob/develop/website/src/controllers/groups.js#L237-L246

I wrote the tests as if we were sending an object, with message being one of the keys (like we do with PMs), but it actually just gets sent as a query in the url. I'm working to correct the tests now.

**Comment by crookedneighbor** [comment]
Changed the error to 400, changed the wording to say "You cannot send a blank message". @Horacious notating partial credit for this since we collaborated on it.

**Pull request merged**